packages/repo-metadata-lint/src/get-repo-files.ts

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {OctokitType} from './utils/octokit-util';
+import {Storage} from '@google-cloud/storage';
+import {logger} from 'gcf-utils';
+
+/**
+ * This interface is used for comforting typescript's type system to
+ * deal with the response from `octokit.repos.getContent`.
+ */
+interface File {
+  content: string | undefined;
+}
+
+/**
+ * This function is used for comforting typescript's type system to
+ * deal with the response from `octokit.repos.getContent`.
+ */
+function isFile(file: File | unknown): file is File {
+  return (file as File).content !== undefined;
+}
+
+// Load full contents of GitHub repository from commit cache,
+// use this to iterate over individual .repo-metadata.json files.
+// This approach is used for the benefit of mono-repos.
+export class GetRepoFiles {
+  octokit: OctokitType;
+  owner: string;
+  repo: string;
+  bucket: string;
+  storage: Storage;
+  constructor(
+    owner: string,
+    repo: string,
+    bucket: string,
+    octokit: OctokitType
+  ) {
+    this.octokit = octokit;
+    this.owner = owner;
+    this.repo = repo;
+    this.bucket = bucket;
+    this.storage = new Storage();
+  }
+  // Emit any .repo-metadata.json files found in file listing.
+  async *repoMetadata() {
+    const files = await this.getFileListing();
+    if (!files) {
+      logger.warn(`unable to find file listing for ${this.owner}/${this.repo}`);
+      return;
+    }
+    for (const path of files.trim().split(/\r?\n/)) {
+      if (path.toLowerCase().endsWith('.repo-metadata.json')) {
+        const resp = (
+          await this.octokit.rest.repos.getContent({
+            owner: this.owner,
+            repo: this.repo,
+            path,
+          })
+        ).data;
+        if (isFile(resp)) {
+          yield [path, Buffer.from(resp.content, 'base64').toString('utf8')];
```

Comment:
So yes, generators are neat.  Are we gaining an actual advantage by using them here?

Reply:
On a repository like `google-cloud-dotnet`, there are 100+ `.repo-metadata.json` files that need to be processed, so being able to write code like this:

```js
for async (const metadata of files.getRepoMetadata\()) {
  // do something with metadata.
}
```

I feel is the right pattern ... I don't see a reason not to use a generator.

Reply:
In general \i find auto-pagination a bit magical, I like that we've been moving towards generators for our GAPIC APIs too, where you do something like:

```js
for async (const secret of client.listSecrets()) {
 // do something with secret.
}
```

☝️ the generator is just the other side of this interface.